lib/containers/utils.pm

```diff
@@ -41,20 +41,61 @@ sub test_seccomp {
     }
 }
 
-sub basic_container_tests {
-    my %args    = @_;
-    my $runtime = $args{runtime};
-    die "You must define the runtime!" unless $runtime;
+sub check_docker_firewall {
+    my $container_name = 'sut_container';
+    my $running        = script_output qq(docker ps -a -q | wc -l);
+    validate_script_output('ip a s docker0', sub { /state DOWN/ }) if $running != 0;
+    assert_script_run "firewall-cmd --list-all --zone=docker";
+    validate_script_output "firewall-cmd --list-interfaces --zone=docker",  sub { /docker0/ };
+    validate_script_output "firewall-cmd --list-interfaces --zone=trusted", sub { /^\s*$/ };
+    # Rules applied before DOCKER. Default is to listen to all tcp connections
+    # ex. output: "1           0        0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0"
+    validate_script_output "iptables -L DOCKER-USER -nvx --line-numbers", sub { /1.+all.+0\.0\.0\.0\/0\s+0\.0\.0\.0\/0/ };
+    assert_script_run "docker run -id --rm --name $container_name -p 1234:1234 " . registry_url('alpine');
+    my $container_ip = container_ip($container_name, "docker");
+    # Each running container should have added a new entry to the DOCKER zone.
+    # ex. output: "1           0        0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.2           tcp dpt:1234"
+    validate_script_output "iptables -L DOCKER -nvx --line-numbers", sub { /1.+ACCEPT.+!docker0 docker0.+$container_ip\s+tcp dpt:1234/ };
+    assert_script_run "docker kill $container_name ";
+}
+
+sub get_docker_version {
+    my $v = script_output("docker --version");
+    record_info "$v", $v =~ /(\d{2}\.\d{2})/;
+    return $v =~ /(\d{2}\.\d{2})/;
+}
+
+sub check_runtime_version {
+    my ($current, $other) = @_;
+    return check_version($other, $current, qr/\d{2}(?:\.\d+)/);
+}
 
+sub container_ip {
+    my ($container, $runtime) = @_;
+    my $ip = script_output "$runtime inspect $container --format='{{.NetworkSettings.IPAddress}}'";
+    record_info "$ip";
+    return $ip;
+}
+
+sub registry_url {
+    my ($container_name, $version_tag) = @_;
     my $registry = get_var('REGISTRY', 'docker.io');
     # Images from docker.io registry are listed without the 'docker.io/library/'
     # Images from custom registry are listed with the 'server/library/'
     # We also filter images the same way they are listed.
-    my $prefix = ($registry =~ /docker\.io/) ? "" : "$registry/library/";
+    my $repo = ($registry =~ /docker\.io/) ? "" : "$registry/library";
+    return $repo unless $container_name;
```

Comment:
```suggestion
    return $registry unless $container_name;
```
When you don't specify image name you care just about the registry URL.

Reply:
i am not sure about this. this is what i had thought as well in first place but the fact is that this would be useful only in one place in `lib/containers/common.pm` which for simplicity can stay as it is. Now i am thinking if even we ever need this return if it is not used. late me take another look.

Reply:
As you now created this routine it should be used wherever possible and it should be as universal as possible.

Reply:
done